### PR TITLE
Add bloom post processing effect

### LINKED ISSUE #168: Add bloom post-processing effect
https://learnopengl.com/Advanced-Lighting/Bloom

~~This implies materials having emissive/glow properties.~~

EDIT: No longer requires a specific property; color values aren't clamped to [0,1], this can be used to infer brightness.

### PR DESCRIPTION
**Pull request by Ev1lbl0w**
Implements a "bloom" effect, to begin adding post-processing capabilities to the engine.

Closes #168.

~~Depends on #161 to be merged.~~

**Update (08/07/2022)** - After initial feedback, the bloom method was changed from a gaussian blur *(implemented from https://learnopengl.com/Advanced-Lighting/Bloom)* to a downsampling and upsampling method *(implemented from https://catlikecoding.com/unity/tutorials/advanced-rendering/bloom/)*, not only creating a much stronger and customizable effect but also achieving lower frame times:
image

### CODE REVIEW

**Comment by Ev1lbl0w** [comment]
Ready for review :+1:

File: engine/CMakeLists.txt [hunk 1]
```diff
@@ -16,6 +16,8 @@ set(CUBOS_ENGINE_SOURCE
     "src/cubos/engine/gl/frame.cpp"
     "src/cubos/engine/gl/renderer.cpp"
     "src/cubos/engine/gl/deferred/renderer.cpp"
+    "src/cubos/engine/gl/pps/bloom.cpp"
+    "src/cubos/engine/gl/pps/copy_pass.cpp"
```
Comment (RiscadoA) [inline engine/CMakeLists.txt:20]: Copy pass is repeated

File: engine/CMakeLists.txt [hunk 2]
```diff
@@ -41,6 +43,8 @@ set(CUBOS_ENGINE_INCLUDE
     "include/cubos/engine/gl/frame.hpp"
     "include/cubos/engine/gl/renderer.hpp"
     "include/cubos/engine/gl/deferred/renderer.hpp"
+    "include/cubos/engine/gl/pps/bloom.hpp"
+    "include/cubos/engine/gl/pps/copy_pass.hpp"
     "include/cubos/engine/gl/pps/manager.hpp"
     "include/cubos/engine/gl/pps/pass.hpp"
     "include/cubos/engine/gl/pps/copy_pass.hpp"
```
Comment (RiscadoA) [inline engine/CMakeLists.txt:50]: Copy pass is repeated

File: engine/include/cubos/engine/gl/pps/bloom.hpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        virtual void execute(std::map<Input, core::gl::Texture2D>& inputs, core::gl::Texture2D prev,
+                             core::gl::Framebuffer out) const override;
+
+    private:
+        unsigned int iterations; ///< Number of iterations for downscale/upscale operation
+        float threshold;         ///< Brighness threshold for applying the effect on.
+        float softThreshold;     ///< Soft brightness threshold to include some samples under the upper threshold.
+        float intensity;         ///< Effect intensity
+
+    private:
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gl/pps/bloom.hpp:53]: Remove unnecessary `private:` section

File: engine/include/cubos/engine/gl/pps/bloom.hpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        glm::uvec2 size;                    ///< Size of the window.
+        core::gl::VertexArray screenQuadVA; ///< Screen quad VA used to render the output.
+        core::gl::Sampler texSampler;       ///< Sampler to use for required textures.
+        core::gl::BlendState blendState;    ///< Blend state required for upscaling process.
+
+    private:
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gl/pps/bloom.hpp:59]: Remove unnecessary `private:` section

File: engine/include/cubos/engine/gl/pps/bloom.hpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        // Extraction pipeline
+        core::gl::Framebuffer extFB;                       ///< Framebuffer of the extraction step.
+        core::gl::ShaderPipeline extPipeline;              ///< Shader pipeline of the extraction step.
+        core::gl::ShaderBindingPoint extInputTexBP;        ///< Input texture binding point.
+        core::gl::ShaderBindingPoint extThresholdFilterBP; ///< Threshold information binding point.
+        core::gl::Texture2D extTex;                        ///< Result texture from extraction step
+
+    private:
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gl/pps/bloom.hpp:67]: Remove unnecessary `private:` section

File: engine/include/cubos/engine/gl/pps/bloom.hpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
+#ifndef CUBOS_ENGINE_GL_PPS_BLOOM_HPP
+#define CUBOS_ENGINE_GL_PPS_BLOOM_HPP
+
+#include <cubos/core/gl/render_device.hpp>
+
+#include <cubos/engine/gl/pps/pass.hpp>
+
+#include <glm/glm.hpp>
+
```
Comment (RiscadoA) [inline engine/include/cubos/engine/gl/pps/bloom.hpp:9]: You can remove the whitespace between the glm include and the std include, we usually split it into engine, core and external includes.

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Sorry for taking so long to review this.
Good work! Don't forget to format the code.

File: engine/CMakeLists.txt (hunk 1 above)
Comment (Ev1lbl0w) [reply engine/CMakeLists.txt:20]: Fixed

File: engine/CMakeLists.txt (hunk 2 above)
Comment (Ev1lbl0w) [reply engine/CMakeLists.txt]: Fixed

File: engine/include/cubos/engine/gl/pps/bloom.hpp (hunk 3 above)
Comment (Ev1lbl0w) [reply engine/include/cubos/engine/gl/pps/bloom.hpp]: Fixed

File: engine/include/cubos/engine/gl/pps/bloom.hpp (hunk 4 above)
Comment (Ev1lbl0w) [reply engine/include/cubos/engine/gl/pps/bloom.hpp]: Fixed

File: engine/include/cubos/engine/gl/pps/bloom.hpp (hunk 5 above)
Comment (Ev1lbl0w) [reply engine/include/cubos/engine/gl/pps/bloom.hpp]: Fixed

File: engine/include/cubos/engine/gl/pps/bloom.hpp (hunk 6 above)
Comment (Ev1lbl0w) [reply engine/include/cubos/engine/gl/pps/bloom.hpp]: Fixed

**Review by RiscadoA — APPROVED** [review]

**Pull request merged**